master/buildbot/process/buildrequest.py

```diff
@@ -290,6 +287,22 @@ def canBeCollapsed(master, br1, br2):
             if selfSS['patch'] or otherSS['patch']:
                 defer.returnValue(False)
                 return
+            # get changes & compare
+            selfChanges = yield master.data.get(('sourcestamps', selfSS['ssid'], 'changes'))
```

Comment:
I think this logic will not work because the `if selfSS['revision'] != otherSS['revision']:` will happen before. You should move the revision check after the change check